cloud-storage/cloud_storage.c

```diff
@@ -0,0 +1,887 @@
+
+static struct _operation_queue_item *_operation_queue = NULL;
+static slock_t *_operation_queue_mutex;
+
+static scond_t *_operation_condition;
+static sthread_t *_operation_thread;
+
+static void _sync_files_execute(folder_type_t folder_type);
+
+static void _upload_file_execute(folder_type_t folder_type, char *file_name);
+
+void cloud_storage_item_free(cloud_storage_item_t *item)
+{
+   cloud_storage_item_t *next;
+
+   while (item)
+   {
+      next = item->next;
+
+      free(item->id);
+      free(item->name);
+
+      if (item->item_type == CLOUD_STORAGE_FILE)
+      {
+         if (item->type_data.file.hash_value)
+         {
+            free(item->type_data.file.hash_value);
+         }
+         if (item->type_data.file.download_url)
+         {
+            free(item->type_data.file.download_url);
+         }
+      }
+
+      if (item->item_type == CLOUD_STORAGE_FOLDER && item->type_data.folder.children)
+      {
+         cloud_storage_item_free(item->type_data.folder.children);
+      }
+
+      free(item);
+
+      item = next;
+   }
+}
+
+static void _free_operation_queue(struct _operation_queue_item *queue)
+{
+   struct _operation_queue_item *next_item;
+
+   while (queue)
+   {
+      switch (queue->operation)
+      {
+         case UPLOAD_FILE:
+            if (queue->operation_parameters.upload.local_file)
+            {
+               free(queue->operation_parameters.upload.local_file);
+            }
+            break;
+         case DOWNLOAD_FILE:
+            if (queue->operation_parameters.download.remote_file)
+            {
+               cloud_storage_item_free(queue->operation_parameters.download.remote_file);
+            }
+            break;
+         default:
+            break;
+      }
+
+      next_item = queue->next;
+      free(queue);
+      queue = next_item;
+   }
+}
+
+static void _operation_thread_loop(void *user_data)
+{
+   while (!_shutdown)
+   {
+      struct _operation_queue_item *task;
+
+      slock_lock(_operation_queue_mutex);
+
+      if (!_operation_queue)
+      {
+         scond_wait(_operation_condition, _operation_queue_mutex);
+      }
+
+      if (_shutdown)
+      {
+         slock_unlock(_operation_queue_mutex);
+         continue;
+      }
+
+      task = _operation_queue;
+      if (_operation_queue)
+      {
+         _operation_queue = _operation_queue->next;
+         task->next = NULL;
+      }
+
+      slock_unlock(_operation_queue_mutex);
+
+      if (task)
+      {
+         switch (task->operation)
+         {
+            case SYNC_FILES:
+               _sync_files_execute(task->operation_parameters.sync.folder_type);
+               break;
+            case UPLOAD_FILE:
+               _upload_file_execute(task->operation_parameters.upload.folder_type, task->operation_parameters.upload.local_file);
+               break;
+            case DOWNLOAD_FILE:
+               break;
+         }
+
+         _free_operation_queue(task);
```

Comment:
Is this ok or should it be _free_operation_queue_item ?

Reply:
This worked since **task->next** was set to NULL earlier. The code for **free_operation_queue** and **free_operation_queue_item** needed to be cleaned up though.

Went with **free_operation_queue_item** here since it is just one item.